lib/graphql/schema/input_object.rb

```diff
@@ -184,21 +184,25 @@ def validate_non_null_input(input, ctx)
             end
           end
 
-          visible_arguments_map = warden.arguments(self).reduce({}) { |m, f| m[f.name] = f; m}
-
-          # Items in the input that are unexpected
-          input.each do |name, value|
-            if visible_arguments_map[name].nil?
-              result.add_problem("Field is not defined on #{self.graphql_name}", [name])
+          # Inject missing required arguments
+          required_inputs = self.arguments.reduce({}) do |m, (argument_name, argument)|
+            if !input.key?(argument_name) && argument.type.non_null? && warden.get_argument(self, argument_name)
+              m[argument_name] = nil
             end
+
+            m
           end
 
-          # Items in the input that are expected, but have invalid values
-          visible_arguments_map.map do |name, argument|
-            argument_result = argument.type.validate_input(input[name], ctx)
-            if !argument_result.valid?
-              result.merge_result!(name, argument_result)
+          input.to_h.merge(required_inputs).each do |argument_name, value|
+            argument = warden.get_argument(self, argument_name)
+            # Items in the input that are unexpected
+            unless argument
+              result.add_problem("Field is not defined on #{self.graphql_name}", [argument_name])
+              next
             end
+            # Items in the input that are expected, but have invalid values
+            argument_result = argument.type.validate_input(value, ctx)
```

Comment:
@rmosolgo it's weird to me that we handle missing required inputs by adding them as nil, and then letting the validation fail that nil isn't an accepted value? Could we not just `add_problem` directly when we notice they're missing?

Reply:
Yah I was also not sure about this behaviour, I would assume this should actually be handled by a different part of code. 

We can add a specific problem if we notice they are missing rather than doing what I am currently doing.

Reply:
> adding them as nil

Do you mean, it worked this way in the old code? AFAIK the old code checked each argument and made sure that the provided value was valid, without special nil handling. I see that in the proposed change here, it has to handle `nil` differently.

Reply:
As I understand it the old code was kind of implicit - it checks every visible argument, and if the argument was required but had no value provided, then `argument.type.validate_input` would produce an error.

Now we're being a bit more explicit about generating a list of `missing_required_inputs`. Rather than merging those explicit nils into the input just so we can call `validate_input` on nil to generate the error... I think we could just generate the error *instead* of building `missing_required_inputs` at all?

Reply:
> generate the error instead of building `missing_required_inputs`

Ah, yeah, fine with me. (And I guess it must have worked this way in the past, if it _wasn't_ triggering `.visible?` 🤔 )

Reply:
> generate the error instead of building `missing_required_inputs`

In that case, I'm a bit unsure of how this should interact/overlap with the [existing static validation of input object attributes](https://github.com/rmosolgo/graphql-ruby/blob/26d8a66/lib/graphql/static_validation/rules/required_input_object_attributes_are_present.rb#L48-L49).

A casual reading of that file initially suggested to me that this check would be [performed earlier](https://graphql-ruby.org/queries/phases_of_execution.html), removing the need to check for required + visible + not provided input arguments via `GraphQL::Schema::InputObject.validate_non_null_input`.

On a related note: the following diff produces errors only in tests related to variable coercion, which makes me think that the [passing tests for required input object arguments](https://github.com/rmosolgo/graphql-ruby/blob/26d8a66/spec/graphql/static_validation/rules/required_input_object_attributes_are_present_spec.rb) are hitting only the static validation checks.

<details><summary>diff</summary>
<p>

```diff
diff --git a/lib/graphql/schema/input_object.rb b/lib/graphql/schema/input_object.rb
index c4fb241fd..0a3a4c52d 100644
--- a/lib/graphql/schema/input_object.rb
+++ b/lib/graphql/schema/input_object.rb
@@ -181,16 +181,13 @@ module GraphQL
             end
           end

-          # Inject missing required arguments
-          missing_required_inputs = self.arguments.reduce({}) do |m, (argument_name, argument)|
+          self.arguments.each do |argument_name, argument|
             if !input.key?(argument_name) && argument.type.non_null? && warden.get_argument(self, argument_name)
-              m[argument_name] = nil
+              result.add_problem("TODO on #{self.graphql_name}", [argument_name])
             end
-
-            m
           end

-          input.merge(missing_required_inputs).each do |argument_name, value|
+          input.each do |argument_name, value|
             argument = warden.get_argument(self, argument_name)
             # Items in the input that are unexpected
             unless argument
```

</p>
</details>


<details><summary>test output</summary>
<p>

```
Failure:
GraphQL::Query::Executor::variable coercion::for input objects with unknown keys in value#test_0001_returns a variable validation error [/Users/chris/src/github.com/grcooper/graphql-ruby/spec/graphql/query/executor_spec.rb:329]
Minitest::Assertion: --- expected
+++ actual
@@ -1,11 +1,10 @@
 {"errors"=>
   [{"message"=>
-     "Variable $input of type [DairyProductInput] was provided invalid value for 0.foo (Field is not defined on DairyProductInput), 0.source (Expected value to not be null)",
+     "Variable $input of type [DairyProductInput] was provided invalid value for 0.source (TODO on DairyProductInput), 0.foo (Field is not defined on DairyProductInput)",
     "locations"=>[{"line"=>1, "column"=>10}],
     "extensions"=>
      {"value"=>[{"foo"=>"bar"}],
       "problems"=>
-       [{"path"=>[0, "foo"],
-         "explanation"=>"Field is not defined on DairyProductInput"},
-        {"path"=>[0, "source"],
-         "explanation"=>"Expected value to not be null"}]}}]}
+       [{"path"=>[0, "source"], "explanation"=>"TODO on DairyProductInput"},
+        {"path"=>[0, "foo"],
+         "explanation"=>"Field is not defined on DairyProductInput"}]}}]}


Failure:
GraphQL::Query::Executor::variable coercion::for required input object fields#test_0001_returns a variable validation error [/Users/chris/src/github.com/grcooper/graphql-ruby/spec/graphql/query/executor_spec.rb:306]
Minitest::Assertion: --- expected
+++ actual
@@ -1,9 +1 @@
-{"errors"=>
-  [{"message"=>
-     "Variable $input of type ReplaceValuesInput! was provided invalid value for values (Expected value to not be null)",
-    "locations"=>[{"line"=>1, "column"=>13}],
-    "extensions"=>
-     {"value"=>{},
-      "problems"=>
-       [{"path"=>["values"],
-         "explanation"=>"Expected value to not be null"}]}}]}
+#<GraphQL::Query::Result @query=... @to_h={"errors"=>[{"message"=>"Variable $input of type ReplaceValuesInput! was provided invalid value for values (TODO on ReplaceValuesInput)", "locations"=>[{"line"=>1, "column"=>13}], "extensions"=>{"value"=>{}, "problems"=>[{"path"=>["values"], "explanation"=>"TODO on ReplaceValuesInput"}]}}]}>


1602 tests, 7612 assertions, 2 failures, 0 errors, 0 skips
```

</p>
</details>

Reply:
My original comment here was an architecture/style question, not a behavioural one. It sounds like there's more we need to dig into here, but if we're confident that the current code works correctly could we get that out and then follow up on the other points?